docs/typescript.md

```diff
@@ -1,14 +1,225 @@
+  productId: number;
+  quantity: number;
+  color?: string; // Optional color
+
+// Omit color if quantity is 1
+const singleItemPayload = Omit<CartItem, "color" extends string ? "color" : never>;
+
+// Omit color for all items if quantity is always 1
+const cartPayload: singleItemPayload[] = [];
+```
+
+#### When augmenting an object type with `&`, remember to apply `Omit` to any overlapping property keys.
+
+* `&` does not automatically overwrite overlapping entries like the spread (...) operator does: https://tsplay.dev/NDJ48W
+```typescript
+type Original = {
+  a: string;
+  b: number;
+  c: boolean;
+}
+type Augment = {
+  c: 'hello'
+}
+type NoOmit = Original & Augment // never
+type UseOmit = Omit<Original, 'c'> & Augment // { a: string; b: number; c: 'hello'; }
+```
+
+* One edge case to be especially careful of is if an optional property is being overwritten. Unlike the above case, this will fail silently: https://tsplay.dev/m0OBxN
+```typescript
+type Original = {
+  a: string;
+  b: number;
+  c?: boolean;
+}
+type Augment = {
+  c?: 'hello'
+}
+type NoOmit = Original & Augment // { a: string; b: number; c?: undefined; }
+```
+
+
+#### Don't use `any`
+
+* `any` doesn't actually represent "any type"; if you declare a value as `any`, it tells TypeScript to disable typechecking for that value anywhere it appears. This means the compiler won't warn you about potential type mismatches or errors, leading to hidden issues that might surface later in development or production.
+* `any` is dangerous because it infects all surrounding and downstream code. For instance, if you have a function that is declared to return `any` which actually returns an object, all properties of that object will be `any`.
+* Need to truly represent something as "any value"? Use `unknown` instead.
+
+#### Avoid type assertions, or explain their purpose whenever necessary.
+Type assertions inform TypeScript about a variable's actual type, using the `as` syntax or the angle bracket ``<Type>`` syntax.
+While type assertions offer flexibility, they can bypass TypeScript's safety net. This means the compiler won't catch potential type mismatches, leaving you responsible for ensuring correctness.
+
+
+While it's generally recommended to avoid type assertions in TypeScript, here are two common scenarios where they might be necessary:
+
+1. Using reduce:
+
+Imagine you have an array of User objects and want to find the youngest user using reduce. You start with an empty accumulator `({})`, but the final result should be a User object.
+```typescript
+interface User {
+  name: string;
+  age: number;
+}
+
+const users: User[] = [
+  // ... user data
+];
+
+const youngestUser = users.reduce((acc, user) => {
+  if (!acc.age || user.age < acc.age) {
+    acc = user; // Here, `acc` might be undefined initially.
+  }
+  return acc;
+}, {} as User); // Assert `{}` as `User` to satisfy the return type.
+
+console.log(youngestUser); // { name: "...", age: maxAge }
+```
```

Comment:
There are four different entries providing conflicting guidelines for `reduce` usage. 

This highlighted portion aligns with this entry:
- https://github.com/MetaMask/contributor-docs/pull/80/files#r1506091213 (lines 38-70)

...and directly contradicts these entries:
- https://github.com/MetaMask/contributor-docs/pull/80/files#r1506093304 (lines 253-285)
- https://github.com/MetaMask/contributor-docs/pull/80/files#r1506095370 (lines 412-439)